lib/model/query/datasets.js

```diff
@@ -89,40 +89,51 @@ SELECT "action" FROM ds
 `;
 
 // Creates or merges dataset, properties and field mapping.
-// Expects dataset:Frame auxed with `formDefId`, `schemaId`,
-// and array of field:Frame auxed with `propertyName`
-const createOrMerge = (dataset, fields, publish = false) => ({ one, Actees, Datasets, Projects }) =>
-  Promise.all([
-    Projects.getById(dataset.projectId).then((o) => o.get()),
-    Datasets.get(dataset.projectId, dataset.name)
-  ])
-    .then(([project, datasetOption]) =>
-      (datasetOption.isDefined()
-        ? datasetOption.get().acteeId
-        : Actees.provision('dataset', project).then((actee) => (actee.id))))
-    .then((acteeId) =>
-      one(_createOrMerge(dataset, fields, acteeId, publish))
-        .catch(error => {
-          if (error.constraint === 'ds_property_fields_dspropertyid_formdefid_unique') {
-            throw Problem.user.invalidEntityForm({ reason: 'Multiple Form Fields cannot be saved to a single Dataset Property.' });
-          }
-          throw error;
-        }))
-    .then((result) => Datasets.get(dataset.projectId, dataset.name).then((ds) => ds.get())
-      .then((ds) => ((publish === true)
-        ? Datasets.getProperties(ds.id).then(properties => ({ ...ds, properties }))
-        : ds.with({ action: result.action }))));
-
-createOrMerge.audit = (dataset, _, fields, publish) => (log) =>
+const createOrMerge = (name, projectId, formDefId, schemaId, fields, publish) => async ({ one, Actees, Datasets, Projects }) => {
+  // Fetching project again because new form endpoint has it and update form does not.
+  const project = await Projects.getById(projectId).then((o) => o.get());
+  const existingDataset = await Datasets.get(project.id, name);
+  const acteeId = existingDataset.isDefined()
+    ? existingDataset.get().acteeId
+    : await Actees.provision('dataset', project).then((actee) => (actee.id));
+  const partial = new Dataset({ name, projectId: project.id }, { formDefId });
+
+  // Prepare dataset properties from form fields:
+  let dsPropertyFields = fields.filter((field) => (field.propertyName));
+
+  if (dsPropertyFields.filter((field) => field.propertyName === 'name' || field.propertyName === 'label').length > 0)
+    throw Problem.user.invalidEntityForm({ reason: 'Invalid Dataset property.' });
+
+  dsPropertyFields = dsPropertyFields.map((field) => new Form.Field(field, { propertyName: field.propertyName, schemaId, formDefId }));
+
+  // Insert or update
+  const result = await one(_createOrMerge(partial, dsPropertyFields, acteeId, publish))
+    .catch(error => {
+      if (error.constraint === 'ds_property_fields_dspropertyid_formdefid_unique') {
+        throw Problem.user.invalidEntityForm({ reason: 'Multiple Form Fields cannot be saved to a single Dataset Property.' });
+      }
+      throw error;
+    });
+
+  // Fetch dataset again with properties in order to write audit log
+  const ds = await Datasets.get(project.id, name).then((o) => o.get());
```

Comment:
Thinking about ways to remove extra queries, especially based on discussion from last review. There's probably a way to not refetch this information using `existingDataset` and/or `partial` above.

Reply:
Things needed by the audit function.. just construct an object with these things in it instead of refetching dataset.
- Dataset with actee id
- fields (filtered)
- properties (re-fetch from database)